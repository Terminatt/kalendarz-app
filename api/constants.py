EMAIL_REGEX = r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'
PASSWORD_MIN_LENGHT = 9
GROUPS = ['administrator', 'regular_user']
TOKEN_EXPIRATION = 24 # In Hours