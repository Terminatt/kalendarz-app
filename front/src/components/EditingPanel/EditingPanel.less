@import '~styles/variables.less';

.editing-panel {
    &-additional {
        margin-bottom: @margin-small;
    }
}