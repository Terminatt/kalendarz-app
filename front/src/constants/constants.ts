export const formLayout = {
    labelCol: { span: 24 },
    wrapperCol: { span: 24 },
};

export const errorMessages = {
    USERNAME_TAKEN: 'Użytkownik o podanej nazwie już istnieje',
    EMAIL_TAKEN: 'Użytkownik o podanym emailu już istnieje',
    PASSWORD_TOO_SIMPLE: 'Hasło jest zbyt proste',
};
