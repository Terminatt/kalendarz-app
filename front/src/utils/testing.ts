export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
