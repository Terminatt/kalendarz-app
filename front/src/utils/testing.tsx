import React, { ComponentType } from 'react';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { RootState, rootReducer } from '@store/index';

export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
export function isTestingEnv(): boolean {
    return process.env.JEST_WORKER_ID !== undefined || process.env.NODE_ENV === 'test';
}

export interface CustomOptions extends RenderOptions {
    preloadedState: RootState;
}

function render(
    ui: React.ReactElement,
    options?: CustomOptions,
) {
    const store = configureStore({ reducer: rootReducer, preloadedState: options?.preloadedState });
    function Wrapper({ children }: {children: React.ReactNode}) {
        return <Provider store={store}>{children}</Provider>;
    }

    return { element: rtlRender(ui, { wrapper: Wrapper as ComponentType, ...options }), store };
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
