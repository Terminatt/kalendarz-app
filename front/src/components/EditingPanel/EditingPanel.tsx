import React, { useCallback, useState } from 'react';
import CustomList, { ListWithSearchProps } from '@components/CustomList/CustomList';
import TwoColumnLayout from '@components/TwoColumnLayout/TwoColumnLayout';
import TwoModesForm, { TwoModesFormProps } from '@components/TwoModesForm/TwoModesForm';
import { BaseItem, GenericReactContent, Id } from '@generics/generics';
import { calculatePageOnDelete, joinClassNames } from '@utils/general';
import { ArrowLeftOutlined } from '@ant-design/icons';
import './EditingPanel.less';
import CustomButton from '@components/CustomButton/CustomButton';

export interface EditingPanel<T> {
    listWithSearchProps: Omit<ListWithSearchProps<T>, 'onSelect' | 'onDelete' | 'selectedItem' | 'onPageChange' | 'dataSource'>;
    twoModesFormProps: Omit<TwoModesFormProps<T>, 'selected' | 'onFormSubmit' | 'onModeChange'>;
    formItems: GenericReactContent;
    className?: string;
    dataSource: T[];
    additionalPanelActive?: boolean;
    additionalPanel?: GenericReactContent;
    onAdditionalPanelBack?: () => void;
    onFormSubmit?: (values: T, page: number, id?: Id) => void;
    onDelete?: (item: T, page: number) => void;
    onPageChange?: (page: number) => void;
}

const EditingPanel = <T extends BaseItem, >(props: EditingPanel<T>): React.ReactElement => {
    const {
        listWithSearchProps, twoModesFormProps, formItems, className, dataSource, additionalPanelActive, additionalPanel,
        onAdditionalPanelBack, onDelete, onFormSubmit, onPageChange,
    } = props;
    const [selected, setSelected] = useState<T | null>();
    const [currentPage, setCurrentPage] = useState<number>(1);

    const onListPageChange = useCallback((page: number) => {
        setCurrentPage(page);
        if (!onPageChange) {
            return;
        }
        onPageChange(page);
    }, [onPageChange]);

    const onListItemSelect = useCallback((item: T | null) => {
        setSelected(item);
    }, []);

    const onSubmit = useCallback((values: T) => {
        if (!onFormSubmit) {
            return;
        }
        onFormSubmit(values, currentPage, selected?.id);
    }, [selected, currentPage, onFormSubmit]);

    const onItemDelete = useCallback((item: T) => {
        let calculatedPage = calculatePageOnDelete(dataSource.length, currentPage);

        if (!calculatedPage) {
            calculatedPage = 1;
        }

        setCurrentPage(calculatedPage);
        setSelected(null);

        if (!onDelete) {
            return;
        }

        onDelete(item, calculatedPage);
    }, [currentPage, dataSource, onDelete]);

    const onModeChange = useCallback(() => {
        setSelected(null);
    }, []);

    return (
        <TwoColumnLayout
            className={joinClassNames(['editing-panel', className])}
            left={
                (
                    <CustomList
                        {...listWithSearchProps}
                        onSelect={onListItemSelect}
                        onDelete={onItemDelete}
                        selected={selected}
                        onPageChange={onListPageChange}
                        pageNumber={currentPage}
                        dataSource={dataSource}
                    />
                )
            }
            right={(
                <>
                    {additionalPanelActive ? (
                        <div className="editing-panel-additional">
                            <div className="editing-panel-additional-header">
                                <CustomButton icon={<ArrowLeftOutlined />} size="small" onClick={onAdditionalPanelBack}>Wróć do panelu edycji</CustomButton>
                            </div>
                            {additionalPanel}
                        </div>
                    ) : (
                        <TwoModesForm {...twoModesFormProps} selected={selected} onFormSubmit={onSubmit} onModeChange={onModeChange}>
                            {formItems}
                        </TwoModesForm>
                    )}
                </>
            )}
        />
    );
};

export default EditingPanel;
