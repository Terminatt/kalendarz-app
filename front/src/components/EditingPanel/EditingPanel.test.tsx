import React from 'react';
import renderer from 'react-test-renderer';
import EditingPanel from './EditingPanel';
import { Form, Input } from 'antd';
import { matchMedia } from '@utils/testing';
import { fireEvent, render } from '@testing-library/react';
import { PAGE_SIZE } from '@constants/constants';


let data: {id: number, name: string}[] = []
describe('Editing Panel Component', () => {
    beforeAll(() => {
        data = [
            {
                id: 1,
                name: 'test1',
            },
            {
                id: 2,
                name: 'test2',
            },
            {
                id: 3,
                name: 'test3',
            }
        ]
    })

    beforeEach(() => {
        matchMedia();
    })

    afterEach(() => {
        data = [
            {
                id: 1,
                name: 'test1',
            },
            {
                id: 2,
                name: 'test2',
            },
            {
                id: 3,
                name: 'test3',
            }
        ]
    })
    it('matches the snapshot', () => {
        const tree = renderer
            .create(
                <EditingPanel
                    listWithSearchProps={{
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        editPrimaryBtnText: 'Edit',
                        primaryBtnText: 'Create'
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
            )
            .toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('triggers onItemSelect', async () => {
        const onItemSelect = jest.fn();
        const element = render(
                <EditingPanel
                    onItemSelect={onItemSelect}
                    listWithSearchProps={{
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
        );
        
        const item = await element.findAllByText('test1');
        fireEvent.click(item[0]);

        expect(onItemSelect).toBeCalledTimes(1);
        expect(onItemSelect).toBeCalledWith({id: 1, name: 'test1'});
    });

    it('triggers onDelete', async () => {
        const onDelete = jest.fn();
        const element = render(
                <EditingPanel
                    onDelete={onDelete}
                    listWithSearchProps={{
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
        );
        
        const btns = await element.findAllByText('Usuń')
        fireEvent.click(btns[0]);

        const confirmBtn = await element.findByText('Tak');
        fireEvent.click(confirmBtn);

        expect(onDelete).toBeCalledTimes(1);
        expect(onDelete).toBeCalledWith({id: 1, name: 'test1'}, 1, expect.any(Function));
    });

    it('triggers onPageChange', async () => {
        const onPageChange = jest.fn();
        const element = render(
                <EditingPanel
                    onPageChange={onPageChange}
                    listWithSearchProps={{
                        total: PAGE_SIZE * 2,
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
        );
        

        const btn = await element.findByText(2);
        fireEvent.click(btn);

        expect(onPageChange).toBeCalledTimes(1);
        expect(onPageChange).toBeCalledWith(2);
    });

    it('triggers onPageChange', async () => {
        const onPageChange = jest.fn();
        const element = render(
                <EditingPanel
                    onPageChange={onPageChange}
                    listWithSearchProps={{
                        total: PAGE_SIZE * 2,
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
        );
        

        const btn = await element.findByText(2);
        fireEvent.click(btn);

        expect(onPageChange).toBeCalledTimes(1);
        expect(onPageChange).toBeCalledWith(2);
    });

    it('triggers onAdditionalPanelBack', async () => {
        const onAdditionalPanelBack = jest.fn();
        const btnName = 'Wróć do panelu edycji';
        let panelActive = true;

        const element = render(
                <EditingPanel
                    additionalPanelActive={panelActive}
                    onAdditionalPanelBack={onAdditionalPanelBack}
                    listWithSearchProps={{
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
                );
                
        const btn = await element.findByText(btnName);
        fireEvent.click(btn);
        expect(onAdditionalPanelBack).toBeCalledTimes(1);

        panelActive = false;
                
        element.rerender(
                    <EditingPanel
                    additionalPanelActive={panelActive}
                    onAdditionalPanelBack={onAdditionalPanelBack}
                    listWithSearchProps={{
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
                )

        expect(await element.queryByTestId('additional-panel')).toBeNull();
    });

    it('does not trigger onAdditionalPanelBack when no item is selected', async () => {
        const onAdditionalPanelBtnClick = jest.fn();
        const additionalPanelBtnText = "Additional Btn";

        const element = render(
                <EditingPanel
                    additionalPanelActive
                    onAdditionalBtnClick={onAdditionalPanelBtnClick}
                    additionalPanelBtnText={additionalPanelBtnText}
                    listWithSearchProps={{
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
                );
                
        const btn = await element.findByText(additionalPanelBtnText);
        fireEvent.click(btn);
        expect(onAdditionalPanelBtnClick).toBeCalledTimes(0);
    });

    it('does trigger onAdditionalPanelBack when item is selected', async () => {
        const onAdditionalPanelBtnClick = jest.fn();
        const additionalPanelBtnText = "Additional Btn";

        const element = render(
                <EditingPanel
                    additionalPanelActive
                    onAdditionalBtnClick={onAdditionalPanelBtnClick}
                    additionalPanelBtnText={additionalPanelBtnText}
                    listWithSearchProps={{
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
                );
        const item = await element.findByText("test1");
        fireEvent.click(item);

        const btn = await element.findByText(additionalPanelBtnText);
        fireEvent.click(btn);
        expect(onAdditionalPanelBtnClick).toBeCalledTimes(1);
    });

    it('changes mode of form on item select', async () => {
        const element = render(
                <EditingPanel
                    listWithSearchProps={{
                        renderContent: (item) => (
                            <div key={item.id}>{item.name}</div>
                        )
                    }}
                    twoModesFormProps={{
                        formProps: {},
                        primaryBtnText: 'Create',
                        editPrimaryBtnText: 'Edit',
                    }}
                    formItems={(
                        <Form.Item name="name">
                            <Input />
                        </Form.Item>
                    )}
                    dataSource={data}
                />
                );
        const item = await element.findByText("test1");
        fireEvent.click(item);

        expect(element.queryByText('Edycja')).not.toBeNull();
    });
});
