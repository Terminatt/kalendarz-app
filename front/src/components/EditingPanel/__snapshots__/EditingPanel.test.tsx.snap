// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editing Panel Component matches the snapshot 1`] = `
<div
  className="two-column editing-panel"
>
  <div
    className="two-column-left"
  >
    <div
      className="custom-list"
    >
      <div
        className="custom-list-container custom-list-alone"
      >
        <div
          className="ant-list ant-list-vertical ant-list-split"
        >
          <div
            className="ant-spin-nested-loading"
          >
            <div
              className="ant-spin-container"
            >
              <ul
                className="ant-list-items"
              >
                <li
                  className="ant-list-item ant-list-item-no-flex custom-list-item"
                  onClick={[Function]}
                >
                  <div>
                    test1
                  </div>
                  <ul
                    className="ant-list-item-action"
                  >
                    <li>
                      <button
                        className="ant-btn ant-btn-sm custom-btn custom-btn-delete"
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        onMouseDown={[Function]}
                        onTouchStart={[Function]}
                        type="button"
                      >
                        <span
                          aria-label="delete"
                          className="anticon anticon-delete"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="delete"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                            />
                          </svg>
                        </span>
                        <span>
                          Usuń
                        </span>
                      </button>
                    </li>
                  </ul>
                </li>
                <li
                  className="ant-list-item ant-list-item-no-flex custom-list-item"
                  onClick={[Function]}
                >
                  <div>
                    test2
                  </div>
                  <ul
                    className="ant-list-item-action"
                  >
                    <li>
                      <button
                        className="ant-btn ant-btn-sm custom-btn custom-btn-delete"
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        onMouseDown={[Function]}
                        onTouchStart={[Function]}
                        type="button"
                      >
                        <span
                          aria-label="delete"
                          className="anticon anticon-delete"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="delete"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                            />
                          </svg>
                        </span>
                        <span>
                          Usuń
                        </span>
                      </button>
                    </li>
                  </ul>
                </li>
                <li
                  className="ant-list-item ant-list-item-no-flex custom-list-item"
                  onClick={[Function]}
                >
                  <div>
                    test3
                  </div>
                  <ul
                    className="ant-list-item-action"
                  >
                    <li>
                      <button
                        className="ant-btn ant-btn-sm custom-btn custom-btn-delete"
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        onMouseDown={[Function]}
                        onTouchStart={[Function]}
                        type="button"
                      >
                        <span
                          aria-label="delete"
                          className="anticon anticon-delete"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="delete"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                            />
                          </svg>
                        </span>
                        <span>
                          Usuń
                        </span>
                      </button>
                    </li>
                  </ul>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div />
    </div>
  </div>
  <div
    className="two-column-right"
  >
    <div
      className="two-modes-form"
    >
      <div
        className="two-modes-form-header"
      >
        <div
          className="two-modes-form-header-text"
        >
          <h2>
            Panel Edycji
          </h2>
        </div>
        <div
          className="two-modes-form-header-tag"
        >
          <span
            className="ant-tag ant-tag-has-color"
            style={
              Object {
                "backgroundColor": "#026E05",
              }
            }
          >
            Nowy
          </span>
        </div>
      </div>
      <form
        className="ant-form ant-form-horizontal custom-form"
        onReset={[Function]}
        onSubmit={[Function]}
      >
        <div
          className="ant-row ant-form-item"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-24 ant-form-item-control"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input"
                  id="name"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="custom-form-btns"
        >
          <button
            className="ant-btn custom-btn custom-btn-primary custom-form-btn"
            onClick={[Function]}
            type="submit"
          >
            <span>
              Create
            </span>
          </button>
          <button
            className="ant-btn custom-btn custom-btn-clear custom-form-btn"
            onClick={[Function]}
            type="button"
          >
            <span>
              Wyczyść
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
